README.md
# ManageIQ Messaging Client

[![Gem Version](https://badge.fury.io/rb/manageiq-messaging.svg)](http://badge.fury.io/rb/manageiq-messaging)
[![CI](https://github.com/ManageIQ/manageiq-messaging/actions/workflows/ci.yaml/badge.svg)](https://github.com/ManageIQ/manageiq-messaging/actions/workflows/ci.yaml)
[![Code Climate](https://codeclimate.com/github/ManageIQ/manageiq-messaging.svg)](https://codeclimate.com/github/ManageIQ/manageiq-messaging)
[![Test Coverage](https://codeclimate.com/github/ManageIQ/manageiq-messaging/badges/coverage.svg)](https://codeclimate.com/github/ManageIQ/manageiq-messaging/coverage)

Client library for ManageIQ components to exchange messages through its internal message bus.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'manageiq-messaging'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install manageiq-messaging

## Usage

### Initialize a client

It is not recommended to directly create an actual client through `new` operation. Follow the example by specifying a protocol. This allows to easily switch the underlying messaging system from one type to another. Currently `:Stomp` and `:Kafka` are implemented.

```
  ManageIQ::Messaging.logger = Logger.new(STDOUT)

  client = ManageIQ::Messaging::Client.open(
    :protocol   => 'Stomp',
    :host       => 'localhost',
    :port       => 61616,
    :password   => 'smartvm',
    :username   => 'admin',
    :client_ref => 'generic_1',
    :encoding   => 'json' # default 'yaml'
  )

  # publish or consume messages using the client

  client.close
```

Alternatively, you can pass a block to `.open` without the need to explicitly close the client.

```
  ManageIQ::Messaging::Client.open(
    :protocol   => 'Stomp',
    :host       => 'localhost',
    :port       => 61616,
    :password   => 'smartvm',
    :username   => 'admin',
    :client_ref => 'generic_1'
  ) do |client|
      # do stuff with the client
    end
  end
```

### Publish and subscribe messages

This is the one-to-one publish/subscribe pattern. Multiple subscribers can subscribe to the same queue but only one will consume the message. Subscribers are load balanced.

```
  client.publish_message(
    :service  => 'ems_operation',
    :affinity => 'ems_amazon1',
    :message  => 'power_on',
    :payload  => {
      :ems_ref => 'u987',
      :id      => '123'
    }
  )

  client.subscribe_messages(:service => 'ems_operation', :affinity => 'ems_amazon1', :auto_ack => false) do |messages|
    messages.each do |msg|
      # do stuff with msg.message and msg.payload
      msg.ack
    end
  end

  # You can create a second client instance and call subscribe_messages with
  # the same options. Then both clients will take turns to consume the messages.
```

For better sending performance, you can publish a collection of messages together

```
  msg1 = {:service => 'ems_inventory', :affinity => 'ems1', :message => 'refresh', :payload => 'vm1'}
  msg2 = {:service => 'ems_inventory', :affinity => 'ems1', :massage => 'refresh', :payload => 'vm2')
  client.publish_messages([msg1, msg2])
```

Provide a block if you want `#publish_message` to wait on a response from the subscriber. This feature may not be supported by every underlying messaging system.

```
  client.publish_message(
    :service  => 'ems_operation',
    :affinity => 'ems_amazon1',
    :message  => 'power_on',
    :payload  => {
      :ems_ref => 'u987',
      :id      => '123'
    }
  ) do |result|
    ansible_install_pkg(vm1) if result == 'running'
  end
```

### Publish and subscribe background jobs

Background Job is a special type of message with a known `class_name` and the subscriber knows how to process the message without a user block

```
  client.publish_message(
    :service    => 'generic',
    :class_name => 'MiqTask',
    :message    => 'update_attributes', # method name
    :payload    => {
      :instance_id => 2,
      :args        => [{:status => 'Timeout'}]
    }
  )

  client.subscribe_background_job(:service => 'generic')
```

Provide a call block if you want `#publish_message` to wait on a response from the subscriber.

### Publish and subscribe topics (events)

This is the one-to-many publish/subscribe pattern. Multiple subscribers can subscribe to the same queue and each one will receive the same message.

```
  client.publish_topic(
    :service => 'provider_events',
    :event   => 'powered_on',
    :sender  => 'ems_amazon1', # optional
    :payload => {
      :ems_ref   => 'uid987',
      :timestamp => '1501091391'
    })

  client.subscribe_topic(:service => 'provider_events', :persist_ref => 'automate_1') do |msg|
    # do stuff with msg.sender, msg.message, and msg.payload. sender may be nil if not set by the publisher
  end
```

By default, events are delivered to live subscribers only. Some messaging systems support persistence with options.

### Publish bulk messages to a topic

Often it is more efficient to publish messages in bulk rather than one-at-a-time.  To do this you can pass an array of messages to the `publish_topic` API:

```ruby
  client.publish_topic(
    [
      {:service => 'provider_events', :event => 'powered_off', :payload => {:ems_ref => 'uid987', :timestamp => '1501091391'}},
      {:service => 'provider_events', :event => 'powered_on', :payload => {:ems_ref => 'uid987', :timestamp => '1501091429'}},
    ]
  )
```

### Add your own headers to a message (Queue or Topic)

If you want you can add in your own headers to the send message

```
  client.publish_topic(
    :service => 'provider_events',
    :event   => 'powered_on',
    :headers => {:request_id => "12345"},
    :payload => {:ems_ref => 'uid987'}

  )

  client.subscribe_topic(:service => 'provider_events', :persist_ref => 'automate_1') do |msg|
    puts "Received event #{msg.message} with request-id: #{msg.headers['request_id']}"
  end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/ManageIQ/manageiq-messaging. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGES
= 0.1.0 - 4-Oct-2018
* Initial release

= 0.1.1 - 20-Nov-2018
* By default upon receiving a message or an event, it is automatically acknowledged. But
  the subscriber can decide to turn off the auto ack feature and ack it in the callback block.

= 0.1.2 - 11-Dec-2018
* Allow to set max_bytes to each_batch when subscribe to a Kafka topic.

= 0.1.4 - 3-Apr-2019
* Add an #ack method to a ReceivedMessage to simplify manual acknowledgements
* Allow caller to provide a session_timeout to kafka consumers, default of 30sec

= 0.1.5 - 6-Jun-2019
* Allow caller to provide extra headers to the message

= 0.1.6 - 6-July-2020
* Rescue message body decoding errors. Re-raise errors raised by users code of processing received messages. 

= 1.0.0 - 28-Sep-2020
* Switch to use rdkafka client

= 1.0.1 - 15-Dec-2020
* Allow all kafka options to be passed
* remove sudo:false from .travis.yml
* Use manageiq-style

= 1.0.2 - 4-Jan-2021
* Rails 6.0 Support

= 1.0.3 - 12-May-2021
* Allow bulk publish of messages to a topic

= 1.1.0 - 09-Feb-2022
* Allow rails 6.1


lib/manageiq/messaging/client.rb
module ManageIQ
  module Messaging
    # The abstract client class. It defines methods needed to publish or subscribe messages.
    # It is not recommended to directly create a solid subclass instance. The proper way is
    # to call class method +Client.open+ with desired protocol. For example:
    #
    #   client = ManageIQ::Messaging::Client.open(
    #     :protocol  => 'Stomp',
    #     :host       => 'localhost',
    #     :port       => 61616,
    #     :password   => 'smartvm',
    #     :username   => 'admin',
    #     :client_ref => 'generic_1',
    #     :encoding   => 'json'
    #   )
    #
    # To close the connection one needs to explicitly call +client.close+.
    # Alternatively if a block is given for the +open+ method, the connection will be closed
    # automatically before existing the block. For example:
    #
    #   ManageIQ::Messaging::Client.open(
    #     :protocol   => 'Stomp'
    #     :host       => 'localhost',
    #     :port       => 61616,
    #     :password   => 'smartvm',
    #     :username   => 'admin',
    #     :client_ref => 'generic_1'
    #   ) do |client|
    #     # do stuff with the client
    #     end
    #   end
    class Client
      # Open or create a connection to the message broker.
      # Expected +options+ keys are:
      # * :protocol (Implemented: 'Stomp', 'Kafka'. Default 'Stomp')
      # * :encoding ('yaml' or 'json'. Default 'yaml')
      # Other connection options are underlying messaging system specific.
      #
      # Returns a +Client+ instance if no block is given.
      def self.open(options)
        protocol = options[:protocol] || :Stomp
        client = Object.const_get("ManageIQ::Messaging::#{protocol}::Client").new(options)

        return client unless block_given?

        begin
          yield client
        ensure
          client.close
        end
        nil
      end

      # Publish a message to a queue. The message will be delivered to only one subscriber.
      # Expected keys in +options+ are:
      # * :service    (service and affinity are used to determine the queue name)
      # * :affinity   (optional)
      # * :class_name (optional)
      # * :message    (e.g. method name or message type)
      # * :payload    (message body, a string or an user object that can be serialized)
      # * :sender     (optional, identify the sender)
      # * :headers    (optional, additional headers to add to the message)
      # Other options are underlying messaging system specific.
      #
      # Optionally a call back block can be provided to wait on the consumer to send
      # an acknowledgment. Not every underlying messaging system supports callback.
      # Example:
      #
      #   client.publish_message(
      #     :service  => 'ems_operation',
      #     :affinity => 'ems_amazon1',
      #     :message  => 'power_on',
      #     :payload  => {
      #       :ems_ref => 'u987',
      #       :id      => '123'
      #     }
      #   ) do |result|
      #     ansible_install_pkg(vm1) if result == 'running'
      #   end
      def publish_message(options, &block)
        assert_options(options, [:message, :service])

        publish_message_impl(options, &block)
      end

      # Publish multiple messages to a queue.
      # An aggregate version of +#publish_message+ but for better performance.
      # All messages are sent in a batch. Every element in +messages+ array is
      # an +options+ hash.
      #
      def publish_messages(messages)
        publish_messages_impl(messages)
      end

      # Subscribe to receive messages from a queue.
      # Expected keys in +options+ are:
      # * :service  (service and affinity are used to determine the queue)
      # * :affinity (optional)
      # * :auto_ack (default true, if it is false, client.ack method must be explicitly called)
      # Other options are underlying messaging system specific.
      #
      # A callback block is needed to consume the messages:
      #
      #   client.subscribe_message(options) do |messages|
      #     messages.each do |msg|
      #       # msg is a type of ManageIQ::Messaging::ReceivedMessage
      #       # attributes in msg
      #       msg.sender
      #       msg.message
      #       msg.payload
      #       msg.ack_ref
      #
      #       msg.ack # needed only when options[:auto_ack] is false
      #       # process the message
      #     end
      #   end
      #
      # With the auto_ack option default to true, the message will be automatically
      # acked immediately after the delivery.
      # Some messaging systems allow the subscriber to ack each message in the
      # callback block. The code in the block can decide when to ack according
      # to whether a message can be retried. Ack the message in the beginning of
      # processing if the message is not re-triable; otherwise ack it after the
      # message is proccessed. Any un-acked message will be redelivered to next subscriber
      # AFTER the current subscriber disconnects normally or abnormally (e.g. crashed).
      #
      # To ack a message call +msg.ack+
      def subscribe_messages(options, &block)
        raise "A block is required" unless block_given?
        assert_options(options, [:service])

        subscribe_messages_impl(options, &block)
      end

      # Subscribe to receive from a queue and run each message as a background job.
      # Expected keys in +options+ are:
      # * :service  (service and affinity are used to determine the queue)
      # * :affinity (optional)
      # * :auto_ack (default true, if it is false, client.ack method must be explicitly called)
      # Other options are underlying messaging system specific.
      #
      # This subscriber consumes messages sent through +publish_message+ with required
      # +options+ keys, for example:
      #
      #     client.publish_message(
      #       :service    => 'generic',
      #       :class_name => 'MiqTask',
      #       :message    => 'update_attributes', # method name, for instance method :instance_id is required
      #       :payload    => {
      #         :instance_id => 2, # database id of class instance stored in rails DB
      #         :args        => [{:status => 'Timeout'}] # argument list expected by the method
      #       }
      #     )
      #
      # Background job assumes each job is not re-triable. It is auto-acked as soon as a request
      # is received
      def subscribe_background_job(options)
        assert_options(options, [:service])

        subscribe_background_job_impl(options)
      end

      # Publish a message as a topic. All subscribers will receive a copy of the message.
      # +messages+ can be either a hash or an array of hashes.
      # Expected keys are:
      # * :service (service is used to determine the topic address)
      # * :event   (event name)
      # * :payload (message body, a string or an user object that can be serialized)
      # * :sender  (optional, identify the sender)
      # * :headers (optional, additional headers to add to the message)
      # Other options are underlying messaging system specific.
      #
      def publish_topic(messages)
        messages = Array.wrap(messages)
        messages.each { |msg| assert_options(msg, [:event, :service]) }

        publish_topic_impl(messages)
      end

      # Subscribe to receive topic type messages.
      # Expected keys in +options+ are:
      # * :service (service is used to determine the topic address)
      # Other options are underlying messaging system specific.
      #
      # Some messaging systems allow subscribers to consume events missed during the period when
      # the client is offline when they reconnect. Additional options are needed to turn on
      # this feature.
      #
      # A callback block is needed to consume the topic:
      #
      #   client.subcribe_topic(:service => 'provider_events', :auto_ack => false) do |msg|
      #     # msg is a type of ManageIQ::Messaging::ReceivedMessage
      #     # attributes in msg
      #     msg.sender
      #     msg.message
      #     msg.payload
      #     msg.ack_ref
      #
      #     msg.ack # needed only when options[:auto_ack] is false
      #     # process the message
      #   end
      #
      # With the auto_ack option default to true, the message will be automatically
      # acked immediately after the delivery.
      # Some messaging systems allow the subscriber to ack each message in the
      # callback block. The code in the block can decide when to ack according
      # to whether a message can be retried. Ack the message in the beginning of
      # processing if the message is not re-triable; otherwise ack it after the
      # message is proccessed. Any un-acked message will be redelivered to next subscriber
      # AFTER the current subscriber disconnects normally or abnormally (e.g. crashed).
      #
      # To ack a message call +msg.ack+
      def subscribe_topic(options, &block)
        raise "A block is required" unless block_given?
        assert_options(options, [:service])

        subscribe_topic_impl(options, &block)
      end

      private

      def logger
        ManageIQ::Messaging.logger
      end

      def assert_options(options, keys)
        missing = keys - options.keys
        raise ArgumentError, "options must contain keys #{missing}" unless missing.empty?
      end
    end
  end
end
